File: node/searchsettings.js
/*
 * nodesearch.js
 *
 * file search utility written in node.js
 */

function SearchSettings() {
    var that = this;
    this.DEFAULT_OUT_DIRPATTERNS = [
        new RegExp("\\bCVS"),
        new RegExp("\\.git"),
        new RegExp("\\.svn")]
    this.DEFAULT_OUT_FILEPATTERNS = [new RegExp("^\\.DS_Store$")]
    this.startPath = "";
    this.inExtensions = [];
    this.outExtensions = [];
    this.inDirPatterns = [];
    this.outDirPatterns = this.DEFAULT_OUT_DIRPATTERNS;
    this.inFilePatterns = [];
    this.outFilePatterns = this.DEFAULT_OUT_FILEPATTERNS;
    this.inArchiveFilePatterns = [];
    this.outArchiveFilePatterns = [];
    this.inLinesAfterPatterns = [];
    this.outLinesAfterPatterns = [];
    this.inLinesBeforePatterns = [];
    this.outLinesBeforePatterns = [];
    this.linesAfterToPatterns = [];
    this.linesAfterUntilPatterns = [];
    this.searchPatterns = [];
    this.archivesOnly = false;
    this.debug = false;
    this.doTiming = false;
    this.excludeHidden = true;
    this.firstMatch = false;
    this.linesAfter = 0;
    this.linesBefore = 0;
    this.listDirs = false;
    this.listFiles = false;
    this.listLines = false;
    this.maxLineLength = 150;
    this.multilineSearch = false;
    this.printResults = false;
    this.printUsage = false;
    this.printVersion = false;
    this.recursive = true;
    this.searchArchives = false;
    this.uniqueLines = false;
    this.verbose = false;
    var addExtension = function (ext, arr) {
        arr.push(ext);
    };
    this.addInExtension = function (ext) {
        addExtension(ext, that.inExtensions);
    };
    this.addOutExtension = function (ext) {
        addExtension(ext, that.outExtensions);
    };
    var addPattern = function (pattern, arr) {
        arr.push(new RegExp(pattern));
    };
    this.addInDirPattern = function (pattern) {
        addPattern(pattern, that.inDirPatterns);
    };
    this.addOutDirPattern = function (pattern) {
        addPattern(pattern, that.outDirPatterns);
    };
    this.addInFilePattern = function (pattern) {
        addPattern(pattern, that.inFilePatterns);
    };
    this.addOutFilePattern = function (pattern) {
        addPattern(pattern, that.outFilePatterns);
    };
    this.addSearchPattern = function (pattern) {
        addPattern(pattern, that.searchPatterns);
    };
    this.addInArchiveFilePattern = function (pattern) {
        addPattern(pattern, that.inArchiveFilePatterns);
    };
    this.addOutArchiveFilePattern = function (pattern) {
        addPattern(pattern, that.outArchiveFilePatterns);
    };
    this.addInLinesAfterPattern = function (pattern) {
        addPattern(pattern, that.inLinesAfterPatterns);
    };
    this.addOutLinesAfterPattern = function (pattern) {
        addPattern(pattern, that.outLinesAfterPatterns);
    };
    this.addInLinesBeforePattern = function (pattern) {
        addPattern(pattern, that.inLinesBeforePatterns);
    };
    this.addOutLinesBeforePattern = function (pattern) {
        addPattern(pattern, that.outLinesBeforePatterns);
    };

    this.addLinesAfterToPattern = function (pattern) {
        addPattern(pattern, that.linesAfterToPatterns);
    };
    this.addLinesAfterUntilPattern = function (pattern) {
        addPattern(pattern, that.linesAfterUntilPatterns);
    };

    this.toString = function () {
        var s = 'SearchSettings(startPath="' + that.startPath + '"';
        if (that.inExtensions.length) {
            s = s + ', inExtensions=["' + that.inExtensions.join('","') + '"]';
        }
        if (that.outExtensions.length) {
            s = s + ', outExtensions=["' + that.outExtensions.join('","') + '"]';
        }
        if (that.inDirPatterns.length) {
            s = s + ', inDirPatterns=["' + that.inDirPatterns.join('","') + '"]';
        }
        if (that.outDirPatterns.length) {
            s = s + ', outDirPatterns=["' + that.outDirPatterns.join('","') + '"]';
        }
        if (that.inFilePatterns.length) {
            s = s + ', inFilePatterns=["' + that.inFilePatterns.join('","') + '"]';
        }
        if (that.outFilePatterns.length) {
            s = s + ', outFilePatterns=["' + that.outFilePatterns.join('","') + '"]';
        }
        if (that.inArchiveFilePatterns.length) {
            s = s + ', inArchiveFilePatterns=["' + that.inArchiveFilePatterns.join('","') + '"]';
        }
        if (that.outArchiveFilePatterns.length) {
            s = s + ', outArchiveFilePatterns=["' + that.outArchiveFilePatterns.join('","') + '"]';
        }
        if (that.inLinesAfterPatterns.length) {
            s = s + ', inLinesAfterPatterns=["' + that.inLinesAfterPatterns.join('","') + '"]';
        }
        if (that.outLinesAfterPatterns.length) {
            s = s + ', outLinesAfterPatterns=["' + that.outLinesAfterPatterns.join('","') + '"]';
        }
        if (that.inLinesBeforePatterns.length) {
            s = s + ', inLinesBeforePatterns=["' + that.inLinesBeforePatterns.join('","') + '"]';
        }
        if (that.outLinesBeforePatterns.length) {
            s = s + ', outLinesBeforePatterns=["' + that.outLinesBeforePatterns.join('","') + '"]';
        }
        if (that.linesAfterToPatterns.length) {
            s = s + ', linesAfterToPatterns=["' + that.linesAfterToPatterns.join('","') + '"]';
        }
        if (that.linesAfterUntilPatterns.length) {
            s = s + ', linesAfterUntilPatterns=["' + that.linesAfterUntilPatterns.join('","') + '"]';
        }
        if (that.searchPatterns.length) {
            s = s + ', searchPatterns=["' + that.searchPatterns.join('","') + '"]';
        }
        s = s + ', archivesOnly=' + that.archivesOnly;
        s = s + ', debug=' + that.debug;
        s = s + ', doTiming=' + that.doTiming;
        s = s + ', excludeHidden=' + that.excludeHidden;
        s = s + ', firstMatch=' + that.firstMatch;
        s = s + ', linesAfter=' + that.linesAfter;
        s = s + ', linesBefore=' + that.linesBefore;
        s = s + ', listDirs=' + that.listDirs;
        s = s + ', listFiles=' + that.listFiles;
        s = s + ', listLines=' + that.listLines;
        s = s + ', maxLineLength=' + that.maxLineLength;
        s = s + ', multilineSearch=' + that.multilineSearch;
        s = s + ', printResults=' + that.printResults;
        s = s + ', printVersion=' + that.printVersion;
        s = s + ', recursive=' + that.recursive;
        s = s + ', searchArchives=' + that.searchArchives;
        s = s + ', uniqueLines=' + that.uniqueLines;
        s = s + ', verbose=' + that.verbose;
        s = s + ')';
        return s;
    };
}

exports.SearchSettings = SearchSettings;

File: go/src/elocale.com/clarkcb/xsearch/searchsettings.go
package xsearch

import (
	"bytes"
	"fmt"
	"strings"
)

type SearchSettings struct {
	StartPath               string
	InExtensions            []*string
	OutExtensions           []*string
	InDirPatterns           *SearchPatterns
	OutDirPatterns          *SearchPatterns
	InFilePatterns          *SearchPatterns
	OutFilePatterns         *SearchPatterns
	InArchiveExtensions     []*string
	OutArchiveExtensions    []*string
	InArchiveFilePatterns   *SearchPatterns
	OutArchiveFilePatterns  *SearchPatterns
	InLinesAfterPatterns    *SearchPatterns
	OutLinesAfterPatterns   *SearchPatterns
	InLinesBeforePatterns   *SearchPatterns
	OutLinesBeforePatterns  *SearchPatterns
	LinesAfterToPatterns    *SearchPatterns
	LinesAfterUntilPatterns *SearchPatterns
	SearchPatterns          *SearchPatterns
	ArchivesOnly            bool
	Debug                   bool
	DoTiming                bool
	ExcludeHidden           bool
	FirstMatch              bool
	LinesAfter              int
	LinesBefore             int
	ListDirs                bool
	ListFiles               bool
	ListLines               bool
	MaxLineLength           int
	MultiLineSearch         bool
	PrintResults            bool
	PrintUsage              bool
	PrintVersion            bool
	Recursive               bool
	SearchArchives          bool
	UniqueLines             bool
	Verbose                 bool
}

func GetDefaultSearchSettings() *SearchSettings {
	return &SearchSettings{
		"",                  // StartPath
		[]*string{},         // InExtensions
		[]*string{},         // OutExtensions
		NewSearchPatterns(), // InDirPatterns
		NewSearchPatterns(), // OutDirPatterns
		NewSearchPatterns(), // InFilePatterns
		NewSearchPatterns(), // OutFilePatterns
		[]*string{},         // InArchiveExtensions
		[]*string{},         // OutArchiveExtensions
		NewSearchPatterns(), // InArchiveFilePatterns
		NewSearchPatterns(), // OutArchiveFilePatterns
		NewSearchPatterns(), // InLinesAfterPatterns
		NewSearchPatterns(), // OutLinesAfterPatterns
		NewSearchPatterns(), // InLinesBeforePatterns
		NewSearchPatterns(), // OutLinesBeforePatterns
		NewSearchPatterns(), // LinesAfterToPatterns
		NewSearchPatterns(), // LinesAfterUntilPatterns
		NewSearchPatterns(), // SearchPatterns
		false,               // ArchivesOnly
		false,               // Debug
		false,               // DoTiming
		true,                // ExcludeHidden
		false,               // FirstMatch
		0,                   // LinesAfter
		0,                   // LinesBefore
		false,               // ListDirs
		false,               // ListFiles
		false,               // ListLines
		150,                 // MaxLineLength
		false,               // MultiLineSearch
		true,                // PrintResults
		false,               // PrintUsage
		false,               // PrintVersion
		true,                // Recursive
		false,               // SearchArchives
		false,               // UniqueLines
		false,               // Verbose
	}
}

func (s *SearchSettings) AddInExtension(xs string) {
	for _, x := range strings.Split(xs, ",") {
		ext := strings.ToLower(x)
		s.InExtensions = append(s.InExtensions, &ext)
	}
}

func (s *SearchSettings) AddOutExtension(xs string) {
	for _, x := range strings.Split(xs, ",") {
		ext := strings.ToLower(x)
		s.OutExtensions = append(s.OutExtensions, &ext)
	}
}

func addPattern(p *string, sp *SearchPatterns) {
	sp.AddPattern(p)
}

func (s *SearchSettings) AddInDirPattern(p string) {
	addPattern(&p, s.InDirPatterns)
}

func (s *SearchSettings) AddOutDirPattern(p string) {
	addPattern(&p, s.OutDirPatterns)
}

func (s *SearchSettings) AddInFilePattern(p string) {
	addPattern(&p, s.InFilePatterns)
}

func (s *SearchSettings) AddOutFilePattern(p string) {
	addPattern(&p, s.OutFilePatterns)
}

func (s *SearchSettings) AddInArchiveExtension(xs string) {
	for _, x := range strings.Split(xs, ",") {
		ext := strings.ToLower(x)
		s.InArchiveExtensions = append(s.InArchiveExtensions, &ext)
	}
}

func (s *SearchSettings) AddOutArchiveExtension(xs string) {
	for _, x := range strings.Split(xs, ",") {
		ext := strings.ToLower(x)
		s.OutArchiveExtensions = append(s.OutArchiveExtensions, &ext)
	}
}

func (s *SearchSettings) AddInArchiveFilePattern(p string) {
	addPattern(&p, s.InArchiveFilePatterns)
}

func (s *SearchSettings) AddOutArchiveFilePattern(p string) {
	addPattern(&p, s.OutArchiveFilePatterns)
}

func (s *SearchSettings) AddInLinesBeforePattern(p string) {
	addPattern(&p, s.InLinesBeforePatterns)
}

func (s *SearchSettings) AddOutLinesBeforePattern(p string) {
	addPattern(&p, s.OutLinesBeforePatterns)
}

func (s *SearchSettings) AddInLinesAfterPattern(p string) {
	addPattern(&p, s.InLinesAfterPatterns)
}

func (s *SearchSettings) AddOutLinesAfterPattern(p string) {
	addPattern(&p, s.OutLinesAfterPatterns)
}

func (s *SearchSettings) AddLinesAfterToPattern(p string) {
	addPattern(&p, s.LinesAfterToPatterns)
}

func (s *SearchSettings) AddLinesAfterUntilPattern(p string) {
	addPattern(&p, s.LinesAfterUntilPatterns)
}

func (s *SearchSettings) AddSearchPattern(p string) {
	addPattern(&p, s.SearchPatterns)
}

func addSearchPatternsToBuffer(name string, sp *SearchPatterns, buffer *bytes.Buffer) {
	buffer.WriteString(fmt.Sprintf("%s: [", name))
	for i, r := range sp.patterns {
		if i > 0 {
			buffer.WriteString(",")
		}
		buffer.WriteString(r.String())
	}
	buffer.WriteString("]")
}

func addStringListToBuffer(name string, list []*string, buffer *bytes.Buffer) {
	buffer.WriteString(fmt.Sprintf("%s: [", name))
	elems := []string{}
	for _, l := range list {
		elems = append(elems, *l)
	}
	buffer.WriteString(strings.Join(elems, ","))
	buffer.WriteString("]")
}

func (s *SearchSettings) String() string {
	var buffer bytes.Buffer
	buffer.WriteString("SearchSettings{")
	buffer.WriteString(fmt.Sprintf("StartPath: %s", s.StartPath))
	buffer.WriteString(", ")
	addStringListToBuffer("InExtensions", s.InExtensions, &buffer)
	buffer.WriteString(", ")
	addStringListToBuffer("OutExtensions", s.OutExtensions, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("InDirPatterns", s.InDirPatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("OutDirPatterns", s.OutDirPatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("InFilePatterns", s.InFilePatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("OutFilePatterns", s.OutFilePatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("InArchiveFilePatterns", s.InArchiveFilePatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("OutArchiveFilePatterns", s.OutArchiveFilePatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("InLinesAfterPatterns", s.InLinesAfterPatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("OutLinesAfterPatterns", s.OutLinesAfterPatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("InLinesBeforePatterns", s.InLinesBeforePatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("OutLinesBeforePatterns", s.OutLinesBeforePatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("LinesAfterToPatterns", s.LinesAfterToPatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("LinesAfterUntilPatterns", s.LinesAfterUntilPatterns, &buffer)
	buffer.WriteString(", ")
	addSearchPatternsToBuffer("SearchPatterns", s.SearchPatterns, &buffer)
	buffer.WriteString(fmt.Sprintf(", ArchivesOnly: %t", s.ArchivesOnly))
	buffer.WriteString(fmt.Sprintf(", Debug: %t", s.Debug))
	buffer.WriteString(fmt.Sprintf(", DoTiming: %t", s.DoTiming))
	buffer.WriteString(fmt.Sprintf(", ExcludeHidden: %t", s.ExcludeHidden))
	buffer.WriteString(fmt.Sprintf(", FirstMatch: %t", s.FirstMatch))
	buffer.WriteString(fmt.Sprintf(", LinesAfter: %d", s.LinesAfter))
	buffer.WriteString(fmt.Sprintf(", LinesBefore: %d", s.LinesBefore))
	buffer.WriteString(fmt.Sprintf(", ListDirs: %t", s.ListDirs))
	buffer.WriteString(fmt.Sprintf(", ListFiles: %t", s.ListFiles))
	buffer.WriteString(fmt.Sprintf(", ListLines: %t", s.ListLines))
	buffer.WriteString(fmt.Sprintf(", MaxLineLength: %d", s.MaxLineLength))
	buffer.WriteString(fmt.Sprintf(", MultiLineSearch: %t", s.MultiLineSearch))
	buffer.WriteString(fmt.Sprintf(", PrintResults: %t", s.PrintResults))
	buffer.WriteString(fmt.Sprintf(", PrintUsage: %t", s.PrintUsage))
	buffer.WriteString(fmt.Sprintf(", PrintVersion: %t", s.PrintVersion))
	buffer.WriteString(fmt.Sprintf(", Recursive: %t", s.Recursive))
	buffer.WriteString(fmt.Sprintf(", SearchArchives: %t", s.SearchArchives))
	buffer.WriteString(fmt.Sprintf(", UniqueLines: %t", s.UniqueLines))
	buffer.WriteString(fmt.Sprintf(", Verbose: %t", s.Verbose))
	buffer.WriteString("}")
	return buffer.String()
}
